### PR TITLE
Feat: Support wallet connect

### PR DESCRIPTION
**Pull request by TejasvCypherock**
Closes https://app.clickup.com/t/37308523/CHI-2001

### CODE REVIEW

**Comment by vbcypherock** [comment]
Task linked: [CHI-2001 Wallet Connect Firmware Enabler](https://app.clickup.com/t/37308523/CHI-2001)

File: common/interfaces/user_interface/ui_input_text.c [hunk 1]
```diff
@@ -273,9 +273,10 @@ static void character_event_handler(lv_obj_t* character, const lv_event_t event)
         
         // 10 is added to the current pixel width to change the alignment to RIGHT incase the last added character overlaps
         // 10 is considered an average width of a pixel
+        char first_char_current_text = data->current_text[0];
         if (get_entered_text_px_width() + 10 <= 98 && strnlen(data->entered_text, MAX_ARRAY_SIZE) <= data->max_input_size) {
             if (data->data_type == DATA_TYPE_PIN && strnlen(data->entered_text, MAX_ARRAY_SIZE) < MAX_PIN_SIZE) {
-                strncat(data->entered_text, data->current_text, 2);
+                strncat(data->entered_text, &first_char_current_text, 1);
```
Comment (ujjwal-cyph) [inline common/interfaces/user_interface/ui_input_text.c:279]: put this as a separate change request

File: src/level_four/core/controller/send_transaction_controller_eth.c [hunk 2]
```diff
@@ -57,187 +57,189 @@
...
-    .gas_limit_size = {0}, .gas_limit = {0},
-    .to_address = {0},
-    .value_size = {0}, .value = {0},
-    .payload_size = 0, .payload = NULL,
-    .chain_id_size = {0}, .chain_id = {0},
-    .dummy_r = {0}, .dummy_s = {0},
+    .nonce_size     = {0},
+    .nonce          = {0},
+    .gas_price_size = {0},
+    .gas_price      = {0},
```
Comment (ujjwal-cyph) [inline src/level_four/core/controller/send_transaction_controller_eth.c:77]: this is a big PR. please undo unrelated changes.

File: common/coin_support/near.c [hunk 3]
```diff
@@ -58,128 +58,129 @@
...
-    memzero(path, sizeof(path));
-    memzero(seed, sizeof(seed));
-    memzero(&hdnode, sizeof(hdnode));
+void near_sig_unsigned_byte_array(const uint8_t *unsigned_txn_byte_array,
+                                  uint64_t unsigned_txn_len,
+                                  const txn_metadata *transaction_metadata,
+                                  const char *mnemonics,
+                                  const char *passphrase,
+                                  uint8_t *sig) {
+  uint8_t digest[32] = {0};
```
Comment (ujjwal-cyph) [inline common/coin_support/near.c:118]: please undo unrelated changes.

File: common/interfaces/desktop_app_interface/communication.c [hunk 4]
```diff
@@ -677,17 +677,20 @@ static comm_error_code_t comm_process_cmd_packet(const packet_t *rx_packet) {
...
-    if (rx_packet->header.total_chunks != 1) return INVALID_CHUNK_COUNT;
-    if (rx_packet->header.payload_length != 6) return INVALID_PAYLOAD_LENGTH;
+    if (comm_status.curr_cmd_seq_no != rx_packet->header.sequence_no)
+        return INVALID_SEQUENCE_NO;
+    if (rx_packet->header.chunk_number != 1)
+        return INVALID_CHUNK_NO;
+    if (rx_packet->header.total_chunks != 1)
+        return INVALID_CHUNK_COUNT;
+    if (rx_packet->header.payload_length != 6)
+        return INVALID_PAYLOAD_LENGTH;
```
Comment (ujjwal-cyph) [inline common/interfaces/desktop_app_interface/communication.c:687]: please undo unrelated changes.

**Review by ujjwal-cyph — CHANGES_REQUESTED** [review]

File: STM32L486RGTX_FLASH.ld [hunk 5]
```diff
@@ -39,7 +39,7 @@ MEMORY
 {
   RAM    (xrw)    : ORIGIN = 0x20000000,   LENGTH = 96K
 /*  RAM2    (xrw)    : ORIGIN = 0x10000000,   LENGTH = 32K*/
-  FLASH    (rx)    : ORIGIN = 0x8023000,   LENGTH = 850K
+  FLASH    (rx)    : ORIGIN = 0x8023000,   LENGTH = 868K
```
Comment (ujjwal-cyph) [inline STM32L486RGTX_FLASH.ld:42]: not required

File: common/protocol_buffers/pb.h [hunk 6]
```diff
@@ -0,0 +1,926 @@
+/* Common parts of the nanopb library. Most of these are quite low-level
```
Comment (ujjwal-cyph) [inline common/protocol_buffers/pb.h:1]: lets do the following: 
1. move the addition of protobuf as a separate independent PR with following things in mind
    a. first commit should link to the original commit files with a relevant reference of the specific commit
    b. second commit will incorporate all our relevant changes needed for our specific usage
3. move the entire protobuf from ~`common/protocol_buffers/`~ to `common/libraries/nanopb`

File: common/coin_support/eth_sign_data/simple.pb.h [hunk 7]
```diff
@@ -0,0 +1,214 @@
+/* Automatically generated nanopb header */
```
Comment (ujjwal-cyph) [inline common/coin_support/eth_sign_data/simple.pb.h:1]: please appropriately rename this file

suggestion(s):
1. `eip712.pb`

File: .clang-format [hunk 8]
```diff
@@ -4,4 +4,5 @@ PointerAlignment: Right
 ColumnLimit: 120
 AlignConsecutiveAssignments: true
 AlignConsecutiveMacros: true
-ReflowComments: false
\ No newline at end of file
+ReflowComments: false
+IndentWidth: 4
```
Comment (ujjwal-cyph) [inline .clang-format:8]: better to take this change as a separate or at the least, a separate commit

**Review by ujjwal-cyph — CHANGES_REQUESTED** [review]

File: src/level_four/core/controller/sign_message_controller_eth.c [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+                    reset_flow_level();
+                }
+
+                eth_init_display_nodes(&current_display_node, &msg_data);
+                ASSERT(current_display_node != NULL);
+            }
+        } break;
+
+        case SIGN_MSG_RAW_MSG_RECEIVED_ETH: {
+            if (is_token_whitelisted(&var_send_transaction_data.transaction_metadata))
```
[the remark this answers is not in the document]
Comment (TejasvCypherock) [reply src/level_four/core/controller/sign_message_controller_eth.c:108]: This case parallels SEND_TXN_UNSIGNED_TXN_RECEIVED_ETH in send flow.

File: common/coin_support/eth.c [hunk 10]
```diff
@@ -305,18 +348,188 @@ bool eth_validate_unsigned_txn(eth_unsigned_txn *eth_utxn_ptr, txn_metadata *met
...
+                                       ETHEREUM_ADDRESS_LENGTH) == 0);
+                    }
+                }
+                break;
+            default:
+                return (ETH_ExtractArguments(eth_utxn_ptr->payload, eth_utxn_ptr->payload_size) ==
+                        ETH_UTXN_ABI_DECODE_OK);
+        }
+    } else {
+        return false;
```
[the remark this answers is not in the document]
Comment (TejasvCypherock) [reply common/coin_support/eth.c:370]: No, it won't. This else condition will only run when the payload size is greater than 0 and we do not support it.

File: common/coin_support/eth.c [hunk 11]
```diff
@@ -305,18 +348,188 @@ bool eth_validate_unsigned_txn(eth_unsigned_txn *eth_utxn_ptr, txn_metadata *met
...
+                return true;
             }
         }
+        return false;
     }
+    return true;
+}
 
-  return true;
+bool get_data_as_string(const MessageData *msg_data, char *output, size_t out_length) {
```
[the remark this answers is not in the document]
Comment (TejasvCypherock) [reply common/coin_support/eth.c]: obsolete function used for debugging, removed now.

[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply common/coin_support/eth.c]: header still there. yet to remove

File: common/coin_support/eth.c [hunk 12]
```diff
@@ -419,4 +632,206 @@ void eth_get_fee_string(eth_unsigned_txn *eth_unsigned_txn_ptr, char *fee_decima
...
+ * X1 wallet parser. If a known function is found, a UI element of type
+ * ui_display_node is created. Also, dpAbiTypeArray is updated to point
+ * the argument type list for that function.
+ * 
+ * @param functionTag The function tag found in the EVM transaction payload
+ * @param dpAbiTypeArray Pointer to start of the argument type array for the
+ * identified function.
+ * @return uint8_t The number of arguments in an identified function.
+ */
+static uint8_t ETH_DetectFunction(const uint32_t functionTag, Abi_Type_e const **dpAbiTypeArray) {
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply common/coin_support/eth.c:615]: Suggestion taken

File: common/coin_support/eth.c [hunk 13]
```diff
@@ -419,4 +632,206 @@ void eth_get_fee_string(eth_unsigned_txn *eth_unsigned_txn_ptr, char *fee_decima
...
+    uint8_t numArgsInFunction  = 0;
+    char *EvmFunctionTitle     = NULL;
+    char *EvmFunctionSignature = NULL;
+
+    switch (functionTag) {
+        case EVM_swap_TAG: {
+            numArgsInFunction    = EVM_swap_NUM_ARGS;
+            *(dpAbiTypeArray)    = (Abi_Type_e *)(&(EVM_swapDataType[0]));
+            EvmFunctionTitle     = (char *)EVM_swap_Title;
+            EvmFunctionSignature = (char *)EVM_swap_Signature;
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply common/coin_support/eth.c:629]: Suggestion taken

File: common/libraries/util/utils.c [hunk 14]
```diff
@@ -478,4 +479,22 @@ bool convert_byte_array_to_decimal_string(const uint8_t len,const uint8_t decima
...
+}
+
+uint8_t UTIL_CheckBound(const uint8_t *pBaseAddr,
+                        const uint32_t totalSizeOfChunk,
+                        const uint8_t *pCurrentSrcAddr,
+                        const uint32_t readSize) {
+    uint8_t returnCode = UTIL_OUT_OF_BOUNDS;
+
+    if ((NULL == pBaseAddr) || (NULL == pCurrentSrcAddr) ||
+        ((uint32_t)pCurrentSrcAddr < (uint32_t)pBaseAddr) || (readSize > totalSizeOfChunk)) {
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply common/libraries/util/utils.c:491]: Right, that makes sense. Suggestion taken

File: common/coin_support/eth.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (TejasvCypherock) [reply common/coin_support/eth.c]: removed

File: common/coin_support/eth_sign_data/eip712_utils.h [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+  char *prefix;
+} queue_node;
+
+typedef struct queue {
+  queue_node *front;
+  queue_node *rear;
+  int count;
+} queue;
+
+queue_node *new_queue_node(TypedDataStruct_TypedDataNode *tree_node, const char *prefix) {
```
[the remark this answers is not in the document]
Comment (TejasvCypherock) [reply common/coin_support/eth_sign_data/eip712_utils.h]: updated

File: common/coin_support/eth.c [hunk 16]
```diff
@@ -305,20 +419,86 @@ bool eth_validate_unsigned_txn(eth_unsigned_txn *eth_utxn_ptr, txn_metadata *met
...
+                                       ETHEREUM_ADDRESS_LENGTH) == 0);
+                    }
+                }
+                break;
+            default:
+                return (ETH_ExtractArguments(eth_utxn_ptr->payload, eth_utxn_ptr->payload_size) ==
+                        ETH_UTXN_ABI_DECODE_OK);
         }
+    } else {
+        return false;
```
Comment (ujjwal-cyph) [inline common/coin_support/eth.c:441]:
```suggestion
```

File: common/coin_support/eth.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply common/coin_support/eth.c]: changed

File: common/coin_support/eth_sign_data/simple.pb.c [hunk 17]
```diff
@@ -0,0 +1,29 @@
+/* Automatically generated nanopb constant definitions */
+/* Generated by nanopb-0.4.6 */
```
[the remark this answers is not in the document]
Comment (TejasvCypherock) [reply common/coin_support/eth_sign_data/simple.pb.c]: this will be picked up when we move the existing communication to protobuf

**Review by ujjwal-cyph — APPROVED** [review]

**Pull request merged**